server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -428,4 +475,57 @@ private static void fillDocAndMetaFields(SearchContext context, FieldsVisitor fi
     static boolean hasSequentialDocs(DocIdToIndex[] docs) {
         return docs.length > 0 && docs[docs.length-1].docId - docs[0].docId == docs.length - 1;
     }
+
+    interface Profiler {
```

Comment:
I went with the internal interface and noop implementation to save from a bunch of null checks that'd have made the whole thing even more invasive. I'd be more than happy to get rid of them if anyone has any good ideas.